libkbfs/journal_server.go

```diff
@@ -359,11 +430,49 @@ func (j *JournalServer) JournalStatus(tlfID TlfID) (TLFJournalStatus, error) {
 	return tlfJournal.getJournalStatus()
 }
 
+// shutdownExistingJournals shuts down all write journals, sets the
+// current UID and verifying key to zero, and returns once all
+// shutdowns are complete. It is safe to call multiple times in a row,
+// and once this is called, EnableExistingJournals may be called
+// again.
+func (j *JournalServer) shutdownExistingJournals(ctx context.Context) {
+	j.log.CDebugf(ctx, "Shutting down existing journals")
+
+	j.lock.Lock()
+	defer j.lock.Unlock()
+	for _, tlfJournal := range j.tlfJournals {
+		tlfJournal.shutdown()
+	}
+
+	// Shutting down atomically is more important than respecting
+	// context cancellation.
+	waitCtx := context.Background()
+	for tlfID, tlfJournal := range j.tlfJournals {
+		err := tlfJournal.wait(waitCtx)
```

Comment:
The background worker could receive the shutdown message before it processes the (one) work signal, it will return immediately and never receive the work signal, and it won't have any corresponding context to cancel.